juju/juju
# Upgrade step for bug 1829393 fix

## Description of change

Follow on to PR 10199: upgrade step to return KVM machines with a modification status of Error to Idle.  For each KVM machine, set the modification status to Idle.  This can not be done from the DatabaseMaster as the un-upgraded models will just set the modification status back to error.

This change includes a new facade, UpgradeSteps, to allow for calls to be made via an API if using the APIContext in upgrade.

Resolve TODO in upgrades/context.go to change APIState() to return base.APICaller instead of api.Connect.

Remove unused CharmRevisionerUpdater() and Cleaner() from api.Connection interface.

## QA steps

1. sudo snap refresh juju --channel 2.5/stable
2. /snap/bin/juju bootstrap was upgrade-kvm 
3. /snap/bin/juju deploy ubuntu --to kvm
3. /snap/bin/juju add-unit ubuntu --to lxd:0 -n2
4. sudo snap refresh juju --stable 
5. /snap/bin/juju upgrade-controller
6.  /snap/bin/juju upgrade-juju
5. watch KVM unit's machine go into error state per bug.
4.  juju upgrade-controller --build-agent
3. juju upgrade-juju
4. watch KVM unit's machine return to started state
4. modification status from juju status --format yaml should be idle 

Wait to ensure the machine and unit agents for the containers stay connected to the controller.

## Documentation changes

N/A

## Bug reference

https://bugs.launchpad.net/juju/+bug/1829393
